src/qibo/transpiler/router.py

```diff
@@ -257,6 +257,32 @@ def update(self, swap: tuple):
         ), self._circuit_logical.index(swap[1])
         self._circuit_logical[idx_0], self._circuit_logical[idx_1] = swap[1], swap[0]
 
+    def undo(self, swap: tuple):
+        """Undo the last swap.
+
+        If the last block is not a swap or does not match the swap to undo, an error is raised.
+        Method works in-place.
+
+        Args:
+            swap (tuple): tuple containing the logical qubits to be swapped.
+        """
+        physical_swap = self.logical_to_physical(swap, index=True)
+        last_swap_block = self._routed_blocks.return_last_block()
+        if last_swap_block.gates[0].__class__ != gates.SWAP or sorted(
+            last_swap_block.qubits
+        ) != sorted(physical_swap):
+            raise_error(
+                ConnectivityError,
+                "The last block does not match the swap to undo.",
+            )
+        self._routed_blocks.remove_block(last_swap_block)
+        self._swaps -= 1
+
+        idx_0, idx_1 = self._circuit_logical.index(
+            swap[0]
+        ), self._circuit_logical.index(swap[1])
+        self._circuit_logical[idx_0], self._circuit_logical[idx_1] = swap[1], swap[0]
```

Comment:
This looks ok code-wise, the only thing that seems a bit weird to me is that the `undo` requires for you to pass the `swap` you wish to undo. I mean, since in principle you undo the last one, you wouldn't need to pass explicitely the swap, but just look for the last one added in a routed block, right? And, indeed, here `swap` is only used to check that the last swap you find is the one you expected.

Reply:
I used `swap` to double-check if the undo function works correctly. I'll remove the checking step.

Also, could you take a look at the question I posted above?

Reply:
Sorry for the delay. Double checking could always be useful, indeed, but, if you want to keep it, I wouldn't make it mandatory at least. In the sense that you could have the `swap=None` argument as optional. In any case, this check should be covered in the tests in principle. 
Regarding the question above, instead, I had a look at it last week. Now, the best person to answer this is probably @Simone-Bordoni, but I believe he is currently on a leave. Thus, I will provide you with my guess, which however may be incorrect. By looking at the `update` function https://github.com/qiboteam/qibo/blob/a2f74249c40e1d76eada8b8b795b5155e1ad6374/src/qibo/transpiler/router.py#L241
it seems that you update the `circuit_logical` mapping, not the physical one (`physical_logical`), meaning that in the layout dict, you do not swap the values, but rather the keys.
Therefore you start from:
```
{"q0": 0, "q1": 2, "q2": 1, "q3": 3}
```
after the `(0, 2)` swap you get:
```
{"q2": 0, "q1": 2, "q0": 1, "q3": 3}
```
and after `(1, 2)`:
```
{"q1": 0, "q2": 2, "q0": 1, "q3": 3}
```
This means that you end up with `circuit_logical = [1, 2, 0, 3]`, which is indeed checked in the test:
``` python
assert circuit_map._circuit_logical == [1, 2, 0, 3]
```
I hope this helped, assuming it is correct. In general, though, this small review of the transpiler code that I did to find the answer, made me think that probably, both, the naming scheme could be improved to make it clearer, and some further documentation should be added to the docstrings. Otherwise, everytime we go back to the code, we risk to waste time in re-understanding everything.

Reply:
Thanks for your response. I noticed that while optimizing SABRE and a better naming scheme seems to be needed. I also found and implemented a more efficient way to manage the mapping. I think it would be best to wait until @Simone-Bordoni returns to discuss the new mapping scheme.

Reply:
I am currently on leaves. If you want to implement a new mapping scheme you can do the PR and I will review it when I will be back.